apis/duck/v1/knative_reference.go

```diff
@@ -54,8 +64,25 @@ func (kr *KReference) Validate(ctx context.Context) *apis.FieldError {
 	if kr.Name == "" {
 		errs = errs.Also(apis.ErrMissingField("name"))
 	}
-	if kr.APIVersion == "" {
-		errs = errs.Also(apis.ErrMissingField("apiVersion"))
+	if isKReferenceGroupAllowed(ctx) {
+		if kr.APIVersion == "" && kr.Group == "" {
+			errs = errs.Also(apis.ErrMissingField("apiVersion")).
+				Also(apis.ErrMissingField("group"))
+		}
+		if kr.APIVersion != "" && kr.Group != "" {
+			errs = errs.Also(&apis.FieldError{
+				Message: "both apiVersion and group are specified",
+				Paths:   []string{"apiVersion", "group"},
+				Details: "Only one of them must be specified",
+			})
+		}
+	} else {
+		if kr.Group != "" {
+			errs = errs.Also(apis.ErrDisallowedFields("group"))
```

Comment:
This makes me worried about how to disable the feature if it had previously been enabled, used, and then disabled. It seems like it will brick the cluster.

Reply:
> This makes me worried about how to disable the feature if it had previously been enabled, used, and then disabled. It seems like it will brick the cluster.

It will, although it's a user mistake to do so... I guess it works the same for serving too, eg what happens if the user does the same thing you sad with https://knative.dev/docs/serving/feature-flags/#kubernetes-node-affinity feature? The API server will start rejecting the nodeAffinity field in the validation and will require the user to manually change their resources to use the "stable" apis.

When you're opting in for experimental features you know you can break your cluster, and you should be careful to avoid that. It's the very same concept as when you use a library which is still not stable, you rely on their methods and then the next version the api breaks, it's up to you to fix it.